src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs

```diff
@@ -146,64 +103,65 @@ internal AsymmetricAdapter(SecurityKey key, string algorithm, HashAlgorithm hash
-                    Initialize(rsa, algorithm);
-                    _disposeCryptoOperators = true;
+                    if (securityKey is RsaSecurityKey rsaSecurityKeyFromJsonWebKey)
+                        IntializeUsingRsaSecurityKey(rsaSecurityKeyFromJsonWebKey, algorithm);
+                    else if (securityKey is X509SecurityKey x509SecurityKeyFromJsonWebKey)
+                        InitializeUsingX509SecurityKey(x509SecurityKeyFromJsonWebKey, algorithm, requirePrivateKey);
+                    else if (securityKey is ECDsaSecurityKey edcsaSecurityKeyFromJsonWebKey)
+                        InitializeUsingEcdsaSecurityKey(edcsaSecurityKeyFromJsonWebKey);
+                    else
+                        throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10684, algorithm, key)));
                 }
-                else if (jsonWebKey.Kty == JsonWebAlgorithmsKeyTypes.EllipticCurve)
-                {
-                    ECDsa = ECDsaAdapter.Instance.CreateECDsa(jsonWebKey, requirePrivateKey);
-                    SignatureFunction = SignWithECDsa;
-                    VerifyFunction = VerifyWithECDsa;
-                    _disposeCryptoOperators = true;
-                }
-                else
-                    throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10684, algorithm, key)));
             }
             else if (key is ECDsaSecurityKey ecdsaKey)
             {
-                ECDsa = ecdsaKey.ECDsa;
+                ECDsaSecurityKey = ecdsaKey;
                 SignatureFunction = SignWithECDsa;
                 VerifyFunction = VerifyWithECDsa;
             }
             else
                 throw LogHelper.LogExceptionMessage(new NotSupportedException(LogHelper.FormatInvariant(LogMessages.IDX10684, algorithm, key)));
         }
 
-        private string Algorithm { get; set; }
+        private void IntializeUsingRsaSecurityKey(RsaSecurityKey rsaSecurityKey, string algorithm)
+        {
+            if (rsaSecurityKey.Rsa != null)
+            {
+                InitializeUsingRsa(rsaSecurityKey.Rsa, algorithm);
+            }
+            else
+            {
+                var rsa = RSA.Create();
+                rsa.ImportParameters(rsaSecurityKey.Parameters);
+                InitializeUsingRsa(rsa, algorithm);
+                _disposeCryptoOperators = true;
+            }
+        }
+
+        private void InitializeUsingX509SecurityKey(X509SecurityKey x509SecurityKey, string algorithm, bool requirePrivateKey)
+        {
+            if (requirePrivateKey)
+                InitializeUsingRsa(x509SecurityKey.PrivateKey as RSA, algorithm);
+            else
+                InitializeUsingRsa(x509SecurityKey.PublicKey as RSA, algorithm);
+        }
+
+        private void InitializeUsingEcdsaSecurityKey(ECDsaSecurityKey ecdsaSecurityKey)
+        {
+            ECDsaSecurityKey = ecdsaSecurityKey;
```

Comment:
Do we need to set  `_disposeCryptoOperators = true` here?
And possibly here:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/b553f2db7433ec87fd925147aceb0b83948920f8/src/Microsoft.IdentityModel.Tokens/AsymmetricAdapter.cs#L286-L291

Reply:
I'll check

Reply:
@GeoK I don't think so because we are only creating a single 'derived' key each time because JsonWebKey.ConvertedSecurityKey is set once.

Reply:
@GeoK 	Opened https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1240 to capture this issue.